src/python/pants/engine/internals/parser.py

```diff
@@ -286,12 +287,21 @@ def error_on_imports(build_file_content: str, filepath: str) -> None:
         lineno, _ = token[2]
         if token_str != "import":
             continue
-        raise ParseError(
-            f"Import used in {filepath} at line {lineno}. Import statements are banned in "
-            "BUILD files because they can easily break Pants caching and lead to stale results. "
-            f"\n\nInstead, consider writing a macro ({doc_url('macros')}) or "
-            f"writing a plugin ({doc_url('plugins-overview')}."
-        )
+        filetype = "macro" if os.path.basename(filepath) == "macros.py" else "build"
+        message = {
+            "macro": (
+                f"Import used in {filepath} at line {lineno}. Import statements are banned in "
+                f"macros (that act like a normal BUILD file) because they can easily break Pants caching and lead "
+                f"to stale results. "
+                f"\n\nInstead, consider writing a plugin ({doc_url('plugins-overview')})."
+            ),
+            "build": (
+                f"Import used in {filepath} at line {lineno}. Import statements are banned in "
+                f"BUILD files because they can easily break Pants caching and lead to stale results. "
+                f"\n\nInstead, consider writing a macro ({doc_url('macros')})."
```

Comment:
I think it should be highlighted already for the build file message, that imports will not work when using macros either. The suggestion to use macros is a way to guide users away from the need for using imports at all (which I have a hard time seeing how that is at all helpful, but perhaps another question) which also could be made clear in that hint, I think (to be more helpful than misleading).

Reply:
Totally agree. I've decided to generalize the error message as indeed it may look like writing a macro will help with the imports. Hope that's better now.